### PR TITLE
wasm: support OCI image fetch in istio-agent.

### PR DESCRIPTION
**Pull request by mathetake**
Follow up on #33812. Add support for OCI image fetch in istio-agent for Wasm plugins.

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

**Comment by mathetake** [comment]
cc @bianpengyuan @mandarjog

**Review by dgn — APPROVED** [review]

File: pkg/wasm/cache.go [hunk 1]
```diff
@@ -100,52 +101,63 @@ func NewLocalFileCache(dir string, purgeInterval, moduleExpiry time.Duration) *L
...
+		b, err = fetcher.Fetch(u.Host + u.Path)
+		if err != nil {
+			return "", fmt.Errorf("could not fetch Wasm OCI image: %v", err)
 		}
+	default:
+		return "", fmt.Errorf("unsupported Wasm module downloading URL scheme: %v", u.Scheme)
+	}
 
-		wasmRemoteFetchCount.With(resultTag.Value(fetchSuccess)).Increment()
+	// Get sha256 checksum and check if it is the same as provided one.
```
Comment (dgn) [inline pkg/wasm/cache.go:143]: the API docs currently state that the sha256 field will be used to verify the *container checksum* when using OCI images. we should adhere to that here (or ignore it for now)

Comment (mathetake) [reply pkg/wasm/cache.go:143]: oh is that? what exactly does it mean by container checksum? Manifest's digest? or layer?

Comment (dgn) [reply pkg/wasm/cache.go:143]: container digest == manifest digest afaik

Comment (bianpengyuan) [reply pkg/wasm/cache.go:143]: Should we only do checksum checking for http fetching, and verify fetched image digest instead for OCI?

**Review by bianpengyuan — COMMENTED** [review]
Thanks a lot! lgtm sans the checksum/digest comment.

File: pkg/wasm/cache.go (hunk 1 above)
Comment (mathetake) [reply pkg/wasm/cache.go]: done https://github.com/istio/istio/pull/34430/commits/94714d1908b40ee2d47fea3201f42a90758dc5c2

**Comment by mathetake** [comment]
@bianpengyuan @dgn PTAL.

File: pkg/wasm/cache.go [hunk 2]
```diff
@@ -100,52 +102,69 @@ func NewLocalFileCache(dir string, purgeInterval, moduleExpiry time.Duration) *L
...
 			return "", fmt.Errorf("module downloaded from %v has checksum %v, which does not match: %v", downloadURL, dChecksum, checksum)
 		}
+	case "oci":
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		defer cancel()
+		// TODO: support imagePullSecret and pass it to ImageFetcherOption.
+		fetcher := NewImageFetcher(ctx, ImageFetcherOption{})
+		b, err = fetcher.Fetch(u.Host+u.Path, checksum)
+		if err != nil {
+			return "", fmt.Errorf("could not fetch Wasm OCI image: %v", err)
```
Comment (bianpengyuan) [inline pkg/wasm/cache.go:149]: Can you increment `wasmRemoteFetchCount` counter here with download failure?

Comment (mathetake) [reply pkg/wasm/cache.go:155]: done https://github.com/istio/istio/pull/34430/commits/0d5f0bd97629bc8414eb3d8a967b511d8d3336b1 thanks!

**Review by bianpengyuan — APPROVED** [review]

**Pull request merged**